### PR TITLE
NWChem changes for fftw3 variant

### PR DESCRIPTION
**Pull request by edoapra**
* added fftw3 variant
* patch to fix compilation errors when the fftw3 variant is indeed used https://github.com/nwchemgit/nwchem/issues/792

### CODE REVIEW

File: var/spack/repos/builtin/packages/nwchem/package.py [hunk 1]
```diff
@@ -30,12 +30,14 @@ class Nwchem(Package):
...
     variant("openmp", default=False, description="Enables OpenMP support")
     variant("mpipr", default=False, description="Enables ARMCI with progress rank")
+    variant("fftw3", default=False, description="Link against the FFTW library")
 
     # This patch is for the modification of the build system (e.g. compiler flags) and
     # Fortran syntax to enable the compilation with Fujitsu compilers. The modification
     # will be merged to the next release of NWChem (see https://github.com/nwchemgit/nwchem/issues/347
     # for more detail.
     patch("fj.patch", when="@7.0.2 %fj")
+    patch("fftw_splans.patch")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/nwchem/package.py]: Can you add a brief comment above the `patch` directive explaining:
1. What is the effect of the patch
2. Why it is needed
3. Where it was taken from

?

Comment (alalazo) [reply var/spack/repos/builtin/packages/nwchem/package.py]: Also, the patch is quite big, so I'm wondering if it applies to all versions unconditionally or if it needs to be restricted so some known version range.

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/nwchem/package.py (hunk 1 above)
Comment (edoapra) [reply var/spack/repos/builtin/packages/nwchem/package.py]: * added patch documentation
* patch applies only to version=7.2.0

**Review by alalazo — APPROVED** [review]

File: var/spack/repos/builtin/packages/nwchem/package.py [hunk 2]
```diff
@@ -30,12 +30,23 @@ class Nwchem(Package):
 
     variant("openmp", default=False, description="Enables OpenMP support")
     variant("mpipr", default=False, description="Enables ARMCI with progress rank")
+    variant("fftw3", default=False, description="Link against the FFTW library")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/nwchem/package.py:33]: This variant seems not to be used?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/nwchem/package.py (hunk 2 above)
Comment (edoapra) [reply var/spack/repos/builtin/packages/nwchem/package.py:33]: Have a look at
https://github.com/edoapra/spack/blob/6cbd58bd553c9a99593a80ba7899c2de228b5cba/var/spack/repos/builtin/packages/nwchem/package.py#L116-L119
```
        if "+fftw3" in spec:
            args.extend(["USE_FFTW3=y"])
            args.extend(["LIBFFTW3=%s" % fftw.ld_flags])
            args.extend(["FFTW3_INCLUDE={0}".format(spec["fftw-api"].prefix.include)])
```

Comment (alalazo) [reply var/spack/repos/builtin/packages/nwchem/package.py:33]: Fair, I was only looking at the diff. I guess that means those lines were not hit before.

**Review by alalazo — APPROVED** [review]

File: var/spack/repos/builtin/packages/nwchem/package.py (hunk 2 above)
Comment (edoapra) [reply var/spack/repos/builtin/packages/nwchem/package.py:33]: > Fair, I was only looking at the diff. I guess that means those lines were not hit before.

Yes, that was indeed the case. All this FFTW related bits were introduced a while ago, but they were never active. Once I managed to make them active, that's when the NWChem source bug (that needs to be patched by the big and ugly fftw_splans.patch) was uncovered.

**Pull request merged**